### PR TITLE
DNMY: MOI

### PR DESCRIPTION
**Pull request by odow**
### Blockers
- [x] Decide what to do with LinQuadOptInterface.
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/5
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/6
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/7
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/10
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/17
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/24
- [x] Requires https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/29
- [x] Requires https://github.com/JuliaOpt/MathOptInterface.jl/pull/329
- [x] Requires https://github.com/JuliaOpt/MathOptInterface.jl/pull/354
- [x] Requires https://github.com/JuliaOpt/MathOptInterface.jl/pull/379
- [x] Add MOI (with appropriate tag) to REQUIRE

### This PR

I've copied the code from https://github.com/JuliaOpt/MathOptInterfaceGurobi.jl

<s>MOI tests currently fail.</s>

image

### CODE REVIEW

**Comment by mlubin** [comment]
What about tagging a new release that covers 8.0 before merging this?

**Comment by mlubin** [comment]
Also dropping julia 0.5...

**Comment by odow** [comment]
Yes I was about to ask that. Shall we leave Julia 0.5, tag as v0.3.4, then drop julia 0.5, merge this and tag as v0.4.0?

**Comment by mlubin** [comment]
AFAIK we can't even tag a new version that claims to support 0.5. So drop 0.5 and tag Gurobi.jl 0.4.

**Comment by odow** [comment]
Tests are green!

**Comment by odow** [comment]
Renamed to `MOIWrapper.jl` in `src` and `test`. However I've left the MPB tests in the folder since they are in lots of different files.

**Comment by rdeits** [comment]
I'm getting the same issue here as in https://github.com/JuliaOpt/GLPK.jl/pull/53#issuecomment-388171282 : I can set `@objective model Min x + 1` but not `@objective model Min x` due to `Unsupported attribute MathOptInterface.ObjectiveFunction{MathOptInterface.SingleVariable}()`. Or maybe I'm misunderstanding how the new MOI should be used?

File: test/MOIWrapper.jl [hunk 1]
```diff
@@ -91,3 +91,26 @@ end
...
+        MOI.ScalarAffineFunction{Float64}([x], [1.0], 0.0)
+    )
+
+    cb_calls = Int32[]
+    function callback_function(cb_data::Gurobi.CallbackData, cb_where::Int32)
+        push!(cb_calls, cb_where)
+        nothing
+    end
+
+    MOI.set!(m, Gurobi.CallbackFunction(), callback_function)
```
Comment (odow) [inline test/MOIWrapper.jl]: @joaquimg @mlubin: callbacks in MOI.

File: test/MOIWrapper.jl [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                # (0,2) +---+---+ (2,2)
+                #       |xx/ \xx|
+                #       |x/   \x|
+                #       |/     \|
+                # (0,1) +       + (2,1)
+                #       |       |
+                # (0,0) +---+---+ (2,0)
+                TOL = 1e-6  # Allow for some impreciseness in the solution
+                if y_val - x_val > 1 + TOL
+                    Gurobi.cblazy!(cb_data, m,
```
Comment (odow) [inline test/MOIWrapper.jl]: @mlubin @joaquimg @blegat 

Lazy constraints through MOI

File: test/MOIWrapper.jl [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        #   (2) the location from where the callback was called.
+        # Note that we can access m, x, and y because this function is defined
+        # inside the same scope
+        cb_calls = Int32[]
+        function callback_function(cb_data::Gurobi.CallbackData, cb_where::Int32)
+            push!(cb_calls, cb_where)
+            if cb_where == Gurobi.CB_MIPSOL
+                sol = [0.0, 0.0]
+                Gurobi.cbget_mipsol_sol(cb_data, cb_where, sol)
+                x_val, y_val = sol
```
Comment (odow) [inline test/MOIWrapper.jl]: The only thing is it's not clear how to inject the MIP solution back into MOI-land.

File: test/MOIWrapper.jl (hunk 2 above)
Comment (joaquimg) [reply test/MOIWrapper.jl]: Oarsome!

**Comment by odow** [comment]
This is now rebased on to master.

See JuliaOpt/MathOptInterface.jl#404, JuliaOpt/LinQuadOptInterface.jl#30, JuliaOpt/LinQuadOptInterface.jl#31 for the commented tests

**Comment by tkoolen** [comment]
This (or LQOI) has an issue with duplicate terms in the affine part of the objective function:

```julia
using MathOptInterface
using Gurobi
using Compat.Test

const MOI = MathOptInterface

atol = 1e-4
rtol = 1e-8

# Minimize (x - 1.5)^2 = x^2 - 3 * x + 2.25
model = GurobiOptimizer(OutputFlag=0)
v = MOI.addvariables!(model, 1)

# this works:
# obj = MOI.ScalarQuadraticFunction([MOI.ScalarAffineTerm(-3.0, v[1])], MOI.ScalarQuadraticTerm.(2.0, v, v), 2.25)

# this doesn't:
obj = MOI.ScalarQuadraticFunction([MOI.ScalarAffineTerm(-1.5, v[1]), MOI.ScalarAffineTerm(-1.5, v[1])], MOI.ScalarQuadraticTerm.(2.0, v, v), 2.25)

MOI.set!(model, MOI.ObjectiveFunction{MOI.ScalarQuadraticFunction{Float64}}(), obj)
MOI.set!(model, MOI.ObjectiveSense(), MOI.MinSense)
MOI.optimize!(model)
@test MOI.get(model, MOI.VariablePrimal(), v) ≈ [1.5] atol=atol rtol=rtol
```

(This is the reason for https://github.com/tkoolen/SimpleQP.jl/issues/31#issuecomment-405976579).

File: src/MOIWrapper.jl [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    add_qpterms!(instance.inner, I, J, V)
+    scalediagonal!(V, I, J, 2.0)
+    update_model!(instance.inner)
+end
+
+function LQOI.set_linear_objective!(instance::GurobiOptimizer, columns::Vector{Int}, coefficients::Vector{Float64})
+    nvars = num_vars(instance.inner)
+    obj = zeros(Float64, nvars)
+    for (col, coef) in zip(columns, coefficients)
+        obj[col] = coef
```
Comment (tkoolen) [inline src/MOIWrapper.jl]: This should be a `+=` to support duplicate terms.

**Comment by odow** [comment]
Since MOI and LQOI are settling down, I'm going to merge this. We can fix up any outstanding issues in separate PRs.

**Comment by rdeits** [comment]
@odow did you resolve the issue with duplicate terms? It's fine if not, but we should open an issue to be sure it doesn't get forgotten.

**Pull request merged**